source/endpoints_ruleset.c

```diff
@@ -0,0 +1,939 @@
+    void *user_data) {
+    (void)out_should_continue;
+    struct member_parser_wrapper *wrapper = user_data;
+
+    if (aws_json_value_is_array(value)) {
+        size_t num_elements = aws_json_get_array_size(value);
+        struct aws_array_list *headers = aws_mem_calloc(wrapper->allocator, 1, sizeof(struct aws_array_list));
+        aws_array_list_init_dynamic(headers, wrapper->allocator, num_elements, sizeof(struct aws_string *));
+        if (s_init_array_from_json(wrapper->allocator, value, headers, s_on_header_element)) {
+            AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Failed to extract url.");
+            return aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_PARSE_FAILED);
+        }
+
+        aws_hash_table_put(wrapper->table, aws_string_new_from_cursor(wrapper->allocator, key), headers, NULL);
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+static int s_parse_endpoints_rule_data_endpoint(
+    struct aws_allocator *allocator,
+    const struct aws_json_value *rule_node,
+    struct aws_endpoints_rule_data_endpoint *data_rule) {
+    AWS_PRECONDITION(allocator);
+    AWS_PRECONDITION(rule_node);
+    AWS_PRECONDITION(data_rule);
+
+    data_rule->allocator = allocator;
+    struct aws_json_value *url_node = aws_json_value_get_from_object(rule_node, aws_byte_cursor_from_c_str("url"));
+    if (aws_json_value_is_string(url_node)) {
+        struct aws_byte_cursor url_cur;
+        if (aws_json_value_get_string(url_node, &url_cur)) {
+            AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Failed to extract url.");
+            goto on_error;
+        }
+
+        data_rule->url_type = AWS_ENDPOINTS_URL_TEMPLATE;
+        data_rule->url.template = aws_string_new_from_cursor(allocator, &url_cur);
+    } else {
+        struct aws_string *reference = NULL;
+        if (s_try_parse_reference(allocator, url_node, &reference)) {
+            AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Failed to parse reference.");
+            goto on_error;
+        }
+
+        if (reference != NULL) {
+            data_rule->url_type = AWS_ENDPOINTS_URL_REFERENCE;
+            data_rule->url.reference = reference;
+        } else {
+            data_rule->url_type = AWS_ENDPOINTS_URL_FUNCTION;
+            if (s_parse_function(allocator, url_node, &data_rule->url.function)) {
+                AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Failed to function.");
+                goto on_error;
+            }
+        }
+    }
+
+    struct aws_json_value *properties_node =
+        aws_json_value_get_from_object(rule_node, aws_byte_cursor_from_c_str("properties"));
+    if (properties_node != NULL) {
+
+        struct aws_byte_buf properties_buf;
+        aws_byte_buf_init(&properties_buf, allocator, 0);
+
+        if (aws_byte_buf_append_json_string(properties_node, &properties_buf)) {
+            AWS_LOGF_ERROR(AWS_LS_SDKUTILS_ENDPOINTS_PARSING, "Failed to extract properties.");
+            goto on_error;
```

Comment:
failing to clean up `properties_buf` if things go wrong

use whatever cleanup strategy you want here, it's a weird tmp variable